Resolve symlinked bindmounts

mdcurtis opened the pull request:
This is a more complete (includes tests and handles both Binds and Mounts objects) version of PR#61.

Compared to the original, I've

- Merged the BindsMounts, BindMountsResolved and BindMountsResolvedRO into a single object that's exposed to the policy (BindMounts).  This simplifies the Go code at the potential loss of some convenience in the Rego - however, I think it should be possible to emulate almost all use cases of the separated lists using the new object with some simple Rego
- Refactored the 'extraction (listing) of mounts' into a new function and tested it against a bunch of cases I could think of - in particular, this resolves a bug where an empty Binds list would cause a panic as well as adding (hopefully complete) parsing for Mounts
- Since we now have an object not three lists, non-resolved mounts have "Resolved": "" but with a valid Source - this works for my use case (my current policy just bans all non-resolved mounts) but may not work for everyone.  Open to other options here.

Finally, and this is a limitation of the original PR too - if you want to take advantage of the "Resolved Path" feature, this plugin practically needs to be run as root as a legacy plugin (otherwise it can't resolve all binds).  A workaround, if only a subset of the filesystem needs to be resolved, would be to pass more binds into the managed plugin.  In either case not doing this doesn't cause a bug, it just reduces the usefulness of the symlink resolution feature (which is expected).

Additionally, on some notes on the PR mechanics here, I've added a dependency on testify as it substantially simplifies the unit testing - I've included go.mod/go.sum updates but not pushed updates to the vendor folder - I figure this reduces the noise in the review, but does mean that running "go mod vendor" is required prior to build.  I've not submitted many Go PRs (and certainly none with a vendor folder, so I'm not sure what the expected convention is here).

anderseknert commented on main.go:
```diff
@@ -175,9 +176,56 @@ func (p DockerAuthZPlugin) evaluate(ctx context.Context, r authorization.Request
+	if ok {
+                binds, ok := hostConfig["Binds"].([]interface{})
+                if ok {
+			for _, v := range binds {
```
Indentation looks off here. Could you please gofmt the files included here? Or is this some GitHub display issue?

mdcurtis replied:
Fixed - it was my fault, I wasn't doing this in an environment where gofmt was on by default.

anderseknert replied:
We should probably include a check for that as part of the build steps, but only so much time :)

anderseknert reviewed (commented):
The change looks good to me. The extra dependency seems unnecessary for a single test though. Also, could you please squash the commits and add a signoff for the DCO check please?

anderseknert commented:
Also, perhaps we should add at least something about this in the README, or WDYT?

Either way, thanks for picking it up!

mdcurtis commented:
> 

I've squashed the commits down - I included flixr's commits in that squash as they also weren't signed-off and there was so much rework that I thought that was fair (although I have credited him in the commit message).

Switched from testify to just using reflect, which worked well enough for this case - I'm spoiled by other languages in terms of test frameworks so I normally just turn to testify to have something 'sane' in Go.

Updated README with a bit of documentation, also tried to document some of the other values that aren't directly from the authz messages from docker for completeness.  I'm not really an OPA/Rego expert here so the language may be a bit off.

flixr commented:
nice @mdcurtis !

My rego fu is not strong enough (or it's just too late) to create a policy that checks both the Resolved and readonly properties.
Do you have an example by any chance?
Right now I have:

```
package docker.authz

default allow = false

allow {
    not restricted_host_bind
}

# bind mounts to host with these prefixes are allowed
allowed_host_path_prefixes := {"/allowed/path/prefix/"}

restricted_host_bind
{
    allowed_host_binds := {x | x:= input.BindMounts[_].Resolved; startswith(x, allowed_host_path_prefixes[_])}
    count(input.BindMounts) != count(allowed_host_binds)
}
```
but additonally to the resolved path I only want to allow readonly mounts...
@anderseknert maybe you have a tip on how to do that?
Then we could also add an example to the README.

mdcurtis commented:
Hi @flixr,

I think this would do the job, but in a fairly ugly way (I'm not a Rego expert by any stretch), the following emulates the behaviour in your PR (AFAICT) so that could be a starting point

```

ResolvedMountsRO = output {
  ResolvedMounts := {mount | input.BindMounts[i]; input.BindMounts[i].Resolved != ""; mount := input.BindMounts[i].Resolved }
  ResolvedMountsRO := {mount | input.BindMounts[i]; input.BindMounts[i].ReadOnly; input.BindMounts[i].Resolved != ""; mount := input.BindMounts[i].Resolved }

  print( ResolvedMounts )
  print( ResolvedMountsRO )
  
  output := ResolvedMountsRO
}
```

with a quick test

```
test_resolvedro if {
  ResolvedMountsRO with input as {   "BindMounts": [{"Source":"/etc","ReadOnly":false,"Resolved":"/etc"},{"Source":"/var","ReadOnly":true,"Resolved":"/var"}, {"Source":"/nonexistent","ReadOnly":false,"Resolved":""}] }
}
```

outputs

```
data.docker.authz.test_resolvedro: PASS (774.265µs)

  {"/etc", "/var"}
  {"/var"}

```

flixr commented:
Thanks, using this now:

```
package docker.authz

default allow = false

allow {
    not restricted_host_bind
}

# bind mounts to host with these prefixes are allowed
allowed_host_path_prefixes := {"/allowed/path/prefix/"}

restricted_host_bind
{
    allowed_host_binds := {mount | mount := input.BindMounts[_]; mount.ReadOnly; startswith(mount.Resolved, allowed_host_path_prefixes[_])}
    count(input.BindMounts) != count(allowed_host_binds)
}
```
